Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,647 @@
+//     }
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+)*/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+)?/
+// => `Regex<Substring?>`
+
+// Result builder DSL equivalent:
+//     Optionally {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == Substring?`
+
+/([0-9a-fA-F]+){3}/
+// => `Regex<[Substring]>
+
+// Result builder DSL equivalent:
+//     Repeat(3) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+){3,5}/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat(3...5) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+){3,}/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat(3...) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+```
+
+Note that capturing collections of repeated captures like this is a departure
+from most regular expression implementations, which only provide access to the
+_last_ match of a repeated capture group. For example, Python only captures the
+last group in this dash-separated string:
+
+```python
+rep = re.compile('(?:([0-9a-f]+)-?)+')
+match = rep.match("1234-5678-9abc-def0")
+print(match.group(1))
+# Prints "def0"
+```
+
+By contrast, the proposed Swift version captures all four sub-matches:
+
+```swift
+let pattern = /(?:([0-9a-f]+)-?)+/
+if let match = “1234-5678-9abc-def0".firstMatch(of: pattern) {
+    print(match.captures)
+}
+// Prints [“1234”, “5678", “9abc”, “def0"]
+```
+
+Despite the deviation from prior art, we believe that the proposed capture
+behavior leads to better consistency with the meaning of these quantifiers.
+
+#### Alternation
+
+Alternations are used to match one of multiple possible patterns.
+
+```txt
+pattern ::= pattern '|' pattern
+```
+
+If there are one or more capturing groups within an alternation, the resulting
+capture type is an `Alternation` that's generic over each option's underlying
+pattern.
+
+```swift
+/([01]+)|[0-9]+|([0-9A-F]+)/
+// => `Regex<Alternation<(Substring, Void, Substring)>>`
+```
+
+If there are no capturing groups within an alternation the resulting capture
+type is `Void`.
+
+```swift
+/[01]+|[0-9]+|[0-9A-F]+/
+// => `Regex<Void>`
+```
+
+Nested captures follow the algebra previously described.
+
+```swift
+/([01]+|[0-9]+|[0-9A-F]+)/
+// => `Regex<Substring>`
+/(([01]+)|([0-9]+)|([0-9A-F]+))/
+// => `Regex<(Substring, Alternation<(Substring, Substring, Substring))>>`
+/(?<overall>(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+))/
+// => `Regex<(overall: Substring, Alternation<(binary: Substring, decimal: Substring, hex: Substring))>>`
```

Comment:
I suspect the name `Alternation` won't survive the first 5 posts in the discussion.

Reply:
`Alternation` was used in the result builder syntax in the overview pitch. In that thread, someone suggested `OneOf` instead. Though there's probably needs to be a difference between the type that represents the capture and the type to create the scoped alternation in the result builder.

Reply:
Current I'm distinguishing between the builder type and the match result by scoping: `Alternation` and `RegexMatch.Alternation`.  But I think `OneOf` is a better name for the DSL, so I'm leaving the match result type as `Alternation` for now.